drivers/samsung/displays/mdc_protocol.cr

```diff
@@ -0,0 +1,354 @@
+        switch_to(blanking_input)
+      end
+      do_send(Command::PanelMute, 1)
+    end
+  end
+
+  def hard_off
+    do_send(Command::PanelMute, 0) if self[:power]?.try &.as_bool
+    do_send(Command::HardOff, 0)
+  end
+
+  def power?(**options)
+    do_send(Command::PanelMute, Bytes.empty, **options).get
+    self[:power]
+  end
+
+  # Mutes both audio/video
+  def mute(
+    state : Bool = true,
+    index : Int32 | String = 0,
+    layer : MuteLayer = MuteLayer::AudioVideo
+  )
+  mute_video(state) if layer.video? || layer.audio_video?
+  mute_audio(state) if layer.audio? || layer.audio_video?
+  end
+
+  # Adds video mute state compatible with projectors
+  def mute_video(state : Bool = true)
+    state = state ? 1 : 0
+    do_send(Command::PanelMute, state)
+  end
+
+  # Emulate audio mute
+  def mute_audio(state : Bool = true)
+    # Do nothing if already in desired state
+    return if self[:audio_mute]?.try &.as_bool == state
+    self[:audio_mute] = state
+    if state
+      @previous_volume = self[:volume].as_i
+      volume(0)
+    else
+      volume(@previous_volume)
+    end
+  end
+
+  # check software version
+  def software_version
+    do_send(Command::SoftwareVersion)
+  end
+
+  def serial_number
+    do_send(Command::SerialNumber)
+  end
+
+  def switch_to(input : Input, **options)
+    @input_stable = false
+    input_target = @input_target
+    @input_target = input_target if input_target
+    do_send(Command::Input, input.value, **options)
+  end
+
+  enum SpeakerMode
+    Internal = 0
+    External = 1
+  end
+
+  def speaker_select(mode : SpeakerMode, **options)
+    do_send(Command::Speaker, mode.value, **options)
+  end
+
+  def do_poll
+    do_send(Command::Status, Bytes.empty, priority: 0)
+    power? unless self[:hard_off]?.try &.as_bool
+  end
+
+  DEVICE_SETTINGS = {
+    network_standby: Bool,
+    auto_off_timer: Bool,
+    auto_power: Bool,
+    volume: Int32,
+    contrast: Int32,
+    brightness: Int32,
+    sharpness: Int32,
+    colour: Int32,
+    tint: Int32,
+    red_gain: Int32,
+    green_gain: Int32,
+    blue_gain: Int32
+  }
+  {% for name, kind in DEVICE_SETTINGS %}
+    @[Security(Level::Administrator)]
+    def {{name.id}}(value : {{kind}}, **options)
+      {% if kind.resolve == Bool %}
+        state = value ? 1 : 0
+        data = {{name.id.stringify}} == "auto_off_timer" ? Bytes[0x81, state] : state
+      {% elsif kind.resolve == Int32 %}
+        data = value.clamp(0, 100)
+      {% end %}
+      do_send(Command.parse({{name.id.stringify}}), data, **options)
+    end
+  {% end %}
+
+  def do_device_config
+    {% for name, kind in DEVICE_SETTINGS %}
+      %value = setting?({{kind}}, {{name.id.stringify}})
+      {{name.id}}(%value) unless %value.nil?
+    {% end %}
+  end
+
+  enum ResponseStatus
+    Ack = 0x41 # A
+    Nak = 0x4e # N
+  end
+
+  def received(data, task)
+    hex = data.hexstring
+    logger.debug { "Samsung sent: #{hex}" }
+    data = data.map(&.to_i).to_a
```

Comment:
Is there a reason to remap each of these bytes to an int32?

Reply:
I think I was calculating the checksum incorrectly before and the result of that couldn't be stored in a UInt8 but now that you helped me sort that out, I don't actually need to do this. Thanks for picking this up, I've made this commit.
https://github.com/PlaceOS/drivers/pull/19/commits/886d433be586e9c4c28d9d049937dade3980107c